octobot_evaluators/evaluator/strategy_evaluator.py

```diff
@@ -57,6 +128,201 @@ async def matrix_callback(self,
+        # negative time delta to invalidate evaluations 10 secs before next timestamp
+        nodes_paths_to_authorized_delta = 10
+        # ensure all evaluations are valid (do not trigger on an expired evaluation)
+        if all(is_tentacle_value_valid(self.matrix_id, evaluation_node_path,
+                                       timestamp=current_time,
+                                       delta=nodes_paths_to_authorized_delta)
+               for evaluation_node_path in to_validate_node_paths):
+            self._save_last_evaluation(matrix_id, exchange_name, evaluator_type, evaluator_name,
+                                       cryptocurrency, symbol, time_frame)
+            return True
+        return False
+
+    def _save_last_evaluation(self, matrix_id, exchange_name, evaluator_type, tentacle_name,
+                              cryptocurrency, symbol, time_frame):
+        update_time = get_tentacle_eval_time(matrix_id,
+                                             get_matrix_default_value_path(
+                                                  tentacle_name=tentacle_name,
+                                                  tentacle_type=evaluator_type,
+                                                  exchange_name=exchange_name,
+                                                  cryptocurrency=cryptocurrency,
+                                                  symbol=symbol,
+                                                  time_frame=time_frame)
+                                             )
+        self._set_last_evaluation_time(exchange_name, evaluator_type, cryptocurrency, symbol, time_frame, update_time)
+
+    def _already_sent_this_technical_evaluation(self, matrix_id, evaluator, evaluator_type, exchange_name,
+                                                cryptocurrency, symbol, time_frame):
+        try:
+            update_time = get_tentacle_eval_time(matrix_id,
+                                                 get_matrix_default_value_path(
+                                                      tentacle_name=evaluator,
+                                                      tentacle_type=evaluator_type,
+                                                      exchange_name=exchange_name,
+                                                      cryptocurrency=cryptocurrency,
+                                                      symbol=symbol,
+                                                      time_frame=time_frame)
+                                                 )
+            return self.evaluations_last_updates[exchange_name][evaluator_type][cryptocurrency][symbol][time_frame] \
+                == update_time
+        except KeyError:
+            return False
+
+    def _set_last_evaluation_time(self, exchange_name, evaluator_type, cryptocurrency, symbol, time_frame, value):
+        try:
+            self.evaluations_last_updates[exchange_name][evaluator_type][cryptocurrency][symbol][time_frame] = value
```

Comment:
We'll have a problem here if cryptocurrency is `None` for example.

Reply:
It'll be erased by the next same `evaluator_type` with a `None` cryptocurrency.

Reply:
If 2 symbols are identical yes but in this case it's not a problem for TAs. I added it in case a strategy wants to use it for non-symbol data, in this case they would use only the cryptocurrency part and symbol would be None and it would work the same

Reply:
Ok 👍